socketio/src/asynchronous/client/client.rs

```diff
@@ -401,11 +484,28 @@ impl Client {
         Ok(())
     }
 
+    /// Indicates whether the client should try to reconnect
+    pub(crate) async fn should_reconnect(&self) -> bool {
+        let manually_disconnected = self.manually_disconnected.load(Ordering::Acquire);
+        let server_disconnected = self.server_disconnected.load(Ordering::Acquire);
+
+        if server_disconnected {
+            self.builder.read().await.reconnect_on_disconnect
+        } else {
+            !manually_disconnected
+        }
+    }
+
     /// Returns the packet stream for the client.
     pub(crate) fn as_stream<'a>(
         &'a self,
     ) -> Pin<Box<dyn Stream<Item = Result<Packet>> + Send + 'a>> {
-        stream::unfold(self.socket.clone(), |mut socket| async {
+        let socket_clone = {
+            let s = self.socket.try_read().expect("Socket must be readable");
```

Comment:
Mh, this creates a race condition around the lock... which leads to a panic in the background thread... that's not so nice. Any way we can get rid of this and merge this into the stream logic?

Reply:
Ahh, ok... something like this?

```rust
        let socket_clone = self.socket.clone();

        stream::unfold(socket_clone, |socket| async {
            let mut socket_read = {
                let s = socket.read().await;
                (*s).clone()
            };
```

Reply:
I've updated the PR with this. I think this avoids any chance of a panic.

Reply:
But this means every iteration clones the socket. Maybe I should just do a `blocking_read` outside?

```rust
        let socket_clone = (*self.socket.blocking_read()).clone();

        stream::unfold(socket_clone, |mut socket| async {
            // wait for the next payload
            let packet: Option<std::result::Result<Packet, Error>> = socket.next().await;
```

This is better, right?

Reply:
I prefer the latter! :) `blocking_read` seems like the right thing.

Reply:
This is fixed with the latest commit.